jupyterhub/roles.py

```diff
@@ -0,0 +1,263 @@
+
+    default_roles = [
+        {
+            'name': 'user',
+            'description': 'Everything the user can do',
+            'scopes': ['all'],
+        },
+        {
+            'name': 'admin',
+            'description': 'Admin privileges (currently can do everything)',
+            'scopes': [
+                'all',
+                'users',
+                'users:tokens',
+                'admin:users',
+                'admin:users:servers',
+                'groups',
+                'admin:groups',
+                'read:services',
+                'proxy',
+                'shutdown',
+            ],
+        },
+        {
+            'name': 'server',
+            'description': 'Post activity only',
+            'scopes': ['users:activity!user=username'],
+        },
+    ]
+    return default_roles
+
+
+def get_scopes():
+
+    """
+    Returns a dictionary of scopes:
+    scopes.keys() = scopes of highest level and scopes that have their own subscopes
+    scopes.values() = a list of first level subscopes or None
+    """
+
+    scopes = {
+        'all': ['read:all'],
+        'users': ['read:users', 'users:activity', 'users:servers'],
+        'read:users': [
+            'read:users:name',
+            'read:users:groups',
+            'read:users:activity',
+            'read:users:servers',
+        ],
+        'users:tokens': ['read:users:tokens'],
+        'admin:users': None,
+        'admin:users:servers': None,
+        'groups': ['read:groups'],
+        'admin:groups': None,
+        'read:services': None,
+        'proxy': None,
+        'shutdown': None,
+    }
+
+    return scopes
+
+
+def expand_scope(scopename):
+
+    """Returns a set of all subscopes"""
+
+    scopes = get_scopes()
+    subscopes = [scopename]
+
+    def expand_subscopes(index):
+
+        more_subscopes = list(
+            filter(lambda scope: scope in scopes.keys(), subscopes[index:])
+        )
+        for scope in more_subscopes:
+            subscopes.extend(scopes[scope])
+
+    if scopename in scopes.keys() and scopes[scopename] is not None:
+        subscopes.extend(scopes[scopename])
+        # record the index from where it should check for "subscopes of sub-subscopes"
+        index_for_sssc = len(subscopes)
+        # check for "subscopes of subscopes"
+        expand_subscopes(index=1)
+        # check for "subscopes of sub-subscopes"
+        expand_subscopes(index=index_for_sssc)
+
+    expanded_scope = set(subscopes)
+
+    return expanded_scope
+
+
+def get_subscopes(role=None, roles=None):
+
+    """Returns a set of all available subscopes for a specified role or list of roles"""
+
+    scope_list = []
+    if role:
+        scope_list = role.scopes
+    elif roles:
+        for role in roles:
+            scope_list.extend(role.scopes)
+    else:
+        raise ValueError('Function get_subscopes is missing an argument')
+
+    scopes = list(chain.from_iterable(list(map(expand_scope, scope_list))))
```

Comment:
I think the outer `list` could be changed to `set` and that would save a call in the next line

Reply:
Yes, thank you for spotting this, changed.